File: src/io.c
#include "io.h"

char io_read_port(io_port port)
{
	port_data data;
	__asm__ __volatile__ ("inb %1, %0" : "=a" (data) : "dN" (port));

	return data;
}

void io_write_port(io_port port, port_data data)
{
	__asm__ __volatile__ ("outb %1, %0" : : "dN" (port), "a" (data));
}

File: kernel/src/io.rs
use core::ops::Add;

#[derive(Copy, Clone)]
pub struct Port(pub u16);

impl Port {
    pub unsafe fn write_u8(self, val: u8) {
        asm!("outb %al, %dx" :: "{dx}"(self.0), "{al}"(val));
    }

    pub unsafe fn read_u8(self) -> u8 {
        let ret: u8;
        asm!("inb %dx, %al" : "={ax}"(ret) : "{dx}"(self.0) :: "volatile");
        ret
    }
}

impl Add<u16> for Port {
    type Output = Self;

    fn add(self, rhs: u16) -> Self::Output {
        Self(self.0 + rhs)
    }
}
